src/main/java/net/minestom/server/item/ItemStack.java

```diff
@@ -1,912 +1,177 @@
+    @Contract(pure = true)
+    public boolean isSimilar(@NotNull ItemStack itemStack) {
+        return material.equals(itemStack.material) &&
+                meta.equals(itemStack.meta);
     }
 
     @Override
-    public @NotNull HoverEvent<ShowItem> asHoverEvent(@NotNull UnaryOperator<ShowItem> op) {
-        return HoverEvent.showItem(op.apply(ShowItem.of(this.material, this.amount, NBTUtils.asBinaryTagHolder(this.toNBT().getCompound("tag")))));
-    }
-
-    // Callback events
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
 
-    /**
-     * Called when the player right clicks with this item.
-     *
-     * @param player the player who used the item
-     * @param hand   the hand used
-     */
-    public void onRightClick(@NotNull Player player, @NotNull Player.Hand hand) {
-    }
+        ItemStack itemStack = (ItemStack) o;
+        if (uuid.equals(itemStack.uuid)) return true;
 
-    /**
-     * Called when the player left clicks with this item.
-     *
-     * @param player the player who used the item
-     * @param hand   the hand used
-     */
-    public void onLeftClick(@NotNull Player player, @NotNull Player.Hand hand) {
+        if (amount != itemStack.amount) return false;
+        if (!stackingRule.equals(itemStack.stackingRule)) return false;
+        if (material != itemStack.material) return false;
+        return meta.equals(itemStack.meta);
     }
 
-    /**
-     * Called when the player right clicks with this item on a block.
-     *
-     * @param player    the player who used the item
-     * @param hand      the hand used
-     * @param position  the position of the interacted block
-     * @param blockFace the block face
-     * @return true if it prevents normal item use (placing blocks for instance)
-     */
-    public boolean onUseOnBlock(@NotNull Player player, @NotNull Player.Hand hand, @NotNull BlockPosition position, @NotNull Direction blockFace) {
-        return false;
-    }
-
-    /**
-     * Called when the player click on this item on an inventory.
-     * <p>
-     * Executed before any events.
-     *
-     * @param player          the player who clicked on the item
-     * @param clickType       the click type
-     * @param slot            the slot clicked
-     * @param playerInventory true if the click is in the player inventory
-     */
-    public void onInventoryClick(@NotNull Player player, @NotNull ClickType clickType, int slot, boolean playerInventory) {
-
+    @Override
+    public int hashCode() {
+        int result = stackingRule.hashCode();
+        result = 31 * result + material.hashCode();
+        result = 31 * result + amount;
+        result = 31 * result + meta.hashCode();
+        return result;
+    }
+
+    @Contract(value = "-> new", pure = true)
+    protected @NotNull ItemBuilder builder() {
+        return new ItemBuilder(material, meta.builder())
```

Comment:
Is there a reason we can't clone and store the builder used to create this item initially rather than create a new one? Similar to what is done for meta.

Reply:
I did it this way on meta because there can be an infinite number of fields but in this case only a material & amount so I didn't consider it as necessary, might be wrong

Reply:
Fair enough, I think stylistically it makes sense to do the same for each.